addons/stock/models/stock_move.py

```diff
@@ -2262,44 +2262,78 @@ def _get_upstream_documents_and_responsibles(self, visited):
             return []
 
     def _set_quantity_done_prepare_vals(self, qty):
+        self.ensure_one()
         res = []
+        qty = self.product_uom._compute_quantity(qty, self.product_id.uom_id, round=False)
+        consumed_quant = set()
+        quants = self.env['stock.quant']._get_reserve_quantity(self.product_id, self.location_id, qty)
         for ml in self.move_line_ids:
             ml_qty = ml.quantity
+            if ml.product_uom_id != self.product_id.uom_id:
+                ml_qty = ml.product_uom_id._compute_quantity(ml_qty, self.product_id.uom_id, round=False)
             if self.product_uom.is_zero(qty):
```

Comment:
Since `qty` is expressed in `self.product_id.uom_id`, I guess we need to convert `qty` to the move's uom here?